File: sdks/python/src/verdocs/resources/profiles.py
"""Profile operations (js-sdk: Users/Profiles.ts)."""

from __future__ import annotations

import mimetypes
from os import PathLike
from pathlib import Path
from typing import IO, TYPE_CHECKING, Any

import httpx

from ..errors import VerdocsConnectionError, api_error_from_response
from ..models.core import Profile
from ..models.users import AuthenticateResponse, CreateProfileRequest, UpdateProfileRequest

if TYPE_CHECKING:
    from .._endpoint import AsyncVerdocsEndpoint, VerdocsEndpoint

_PROFILES_PATH = "/v2/profiles"

# The js-sdk allows photo uploads two minutes instead of its default timeout;
# we mirror that for the multipart call.
_PHOTO_TIMEOUT = 120.0

# One multipart file part, in any form httpx files= accepts: raw bytes, a
# binary file-like object, or a (filename, content[, content_type]) tuple.
# str and PathLike values are treated as a filesystem path and read.
FileContent = bytes | IO[bytes]
FileInput = str | PathLike | FileContent | tuple[str | None, FileContent] | tuple[str | None, FileContent, str | None]


def _find_current(payload: list[dict]) -> Profile | None:
    profiles = [Profile.model_validate(entry) for entry in payload]
    return next((profile for profile in profiles if profile.current), None)


def _write_body(params: CreateProfileRequest | UpdateProfileRequest) -> dict[str, Any]:
    # The deployed profile schemas are strict and take optional strings, not
    # nulls, so unset fields stay off the wire and an explicit None does too.
    return params.model_dump(mode="json", exclude_unset=True, exclude_none=True)


def _delete_result(payload: Any) -> AuthenticateResponse | None:
    # Deleting the last profile answers with a logged-out status object
    # instead of tokens; None is our spelling for "no session left".
    if isinstance(payload, dict) and "access_token" in payload:
        return AuthenticateResponse.model_validate(payload)
    return None


def _picture_part(picture: FileInput) -> Any:
    if isinstance(picture, (str, PathLike)):
        path = Path(picture)
        # The server stores the declared part content type with the photo,
        # so we guess one from the file name rather than sending everything
        # as octet-stream.
        return (path.name, path.read_bytes(), mimetypes.guess_type(path.name)[0])
    return picture


class Profiles:
    """Profile calls for a sync endpoint."""

    def __init__(self, endpoint: VerdocsEndpoint) -> None:
        self._endpoint = endpoint

    def list(self) -> list[Profile]:
        """Get the caller's profiles via GET /v2/profiles.

        A user has one profile per organization they are a member of, and
        exactly one is marked current. Mirrors js-sdk getProfiles.

        Returns:
            All of the caller's profiles.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("GET", _PROFILES_PATH)
        return [Profile.model_validate(entry) for entry in response.json()]

    def current(self) -> Profile | None:
        """Get the caller's current profile.

        GET /v2/profiles returns every profile the caller has (one per
        organization); this returns the entry marked current, which is the
        profile all other operations run as. Mirrors js-sdk getCurrentProfile.

        Returns:
            The current profile, or None if no profile is marked current.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("GET", _PROFILES_PATH)
        return _find_current(response.json())

    def create(self, params: CreateProfileRequest) -> AuthenticateResponse:
        """Sign up: create a new user, organization, and profile via POST /v2/profiles.

        This is the self-service registration path, so the deployed endpoint
        requires an UNAUTHENTICATED caller (authenticated users get a 400
        pointing at POST /v2/organizations). The new profile becomes current
        and session tokens are returned, but the session is only partial
        until the emailed verification code is submitted: apply the returned
        access token with set_token(), then call auth.verify_email() on the
        same endpoint. Mirrors js-sdk createProfile.

        Example:
            tokens = endpoint.profiles.create(CreateProfileRequest(
                email="a@b.com", password="secret!A1", first_name="First",
                last_name="Last", org_name="NEW ORG",
            ))
            endpoint.set_token(tokens.access_token)
            verified = endpoint.auth.verify_email(email="a@b.com", token="EMAILED-CODE")

        Args:
            params: The signup fields; org_name does not need to be unique.

        Returns:
            Tokens for the new (partial, unverified) session; apply with set_token().

        Raises:
            VerdocsAPIError: The email already exists, the password is too
                weak, or the caller was already authenticated.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("POST", _PROFILES_PATH, json=_write_body(params))
        return AuthenticateResponse.model_validate(response.json())

    def switch(self, profile_id: str) -> AuthenticateResponse:
        """Switch the caller's current profile via POST /v2/profiles/{profile_id}/switch.

        The current profile drives permissions checking and the profile_id
        stamped on most operations, so switch before acting as another
        organization. Like the js-sdk, the new tokens are returned, not
        applied: call set_token() with the new access token to start
        operating as the switched profile (the old token keeps its previous
        profile context until it expires). Mirrors js-sdk switchProfile.

        Example:
            tokens = endpoint.profiles.switch("PROFILE-ID")
            endpoint.set_token(tokens.access_token)

        Args:
            profile_id: ID of the caller's profile to make current.

        Returns:
            Tokens carrying the switched profile; apply with set_token().

        Raises:
            NotFoundError: The caller has no profile with that ID.
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("POST", f"{_PROFILES_PATH}/{profile_id}/switch")
        return AuthenticateResponse.model_validate(response.json())

    def update(self, profile_id: str, params: UpdateProfileRequest) -> Profile:
        """Update a profile via PATCH /v2/profiles/{profile_id}.

        For your own profile this updates the basic fields (name, phone,
        timezone, locale); admins may also update other members of their
        organization, which is the only path that accepts permissions and
        roles (see UpdateProfileRequest for the split). Mirrors js-sdk
        updateProfile.

        Args:
            profile_id: ID of the profile to update.
            params: The fields to change; unset fields are left alone.

        Returns:
            The updated profile.

        Raises:
            NotFoundError: No profile has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = self._endpoint._request("PATCH", f"{_PROFILES_PATH}/{profile_id}", json=_write_body(params))
        return Profile.model_validate(response.json())

    def update_photo(self, profile_id: str, picture: FileInput) -> Profile:
        """Update the caller's profile photo via a multipart PATCH /v2/profiles/{profile_id}.

        Only the caller's own profile accepts a photo. The file goes up as a
        `picture` part, and the declared content type is stored with it.
        Mirrors js-sdk updateProfilePhoto, minus its browser-only upload
        progress callback.

        Args:
            profile_id: ID of the caller's own profile.
            picture: The photo: a filesystem path (str or PathLike), raw
                bytes, a binary file-like object, or an httpx-style
                (filename, content[, content_type]) tuple.

        Returns:
            The updated profile; its picture field is a public CDN URL.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        path = f"{_PROFILES_PATH}/{profile_id}"
        # _request only speaks JSON, so the multipart upload goes through the
        # raw client with the same error mapping.
        try:
            response = self._endpoint._client.request(
                "PATCH", path, files={"picture": _picture_part(picture)}, timeout=_PHOTO_TIMEOUT
            )
        except httpx.TransportError as exc:
            raise VerdocsConnectionError(f"PATCH {path} failed: {exc}") from exc
        if response.status_code >= 400:
            raise api_error_from_response(response)
        return Profile.model_validate(response.json())

    def delete(self, profile_id: str) -> AuthenticateResponse | None:
        """Delete one of the caller's profiles via DELETE /v2/profiles/{profile_id}.

        Deleting your current profile switches you to the next available one
        and returns its tokens (apply them with set_token()); deleting your
        last remaining profile returns None, and you are logged out. Mirrors
        js-sdk deleteProfile.

        As deployed, the handler only sends a response when the deleted
        profile was the caller's current one: deleting a NON-current profile
        succeeds but never answers, so the call times out. Switch to a
        profile before deleting it.

        Args:
            profile_id: ID of the caller's profile to delete.

        Returns:
            Tokens for the next available profile, or None when the last
            profile was deleted and the session is gone.

        Raises:
            VerdocsAPIError: The profile does not belong to the caller.
            VerdocsConnectionError: The request never reached the API, or
                (deployed quirk above) the API never answered.
        """
        response = self._endpoint._request("DELETE", f"{_PROFILES_PATH}/{profile_id}")
        return _delete_result(response.json())


class AsyncProfiles:
    """Profile calls for an async endpoint."""

    def __init__(self, endpoint: AsyncVerdocsEndpoint) -> None:
        self._endpoint = endpoint

    async def list(self) -> list[Profile]:
        """Get the caller's profiles via GET /v2/profiles.

        A user has one profile per organization they are a member of, and
        exactly one is marked current. Mirrors js-sdk getProfiles.

        Returns:
            All of the caller's profiles.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", _PROFILES_PATH)
        return [Profile.model_validate(entry) for entry in response.json()]

    async def current(self) -> Profile | None:
        """Get the caller's current profile.

        GET /v2/profiles returns every profile the caller has (one per
        organization); this returns the entry marked current, which is the
        profile all other operations run as. Mirrors js-sdk getCurrentProfile.

        Returns:
            The current profile, or None if no profile is marked current.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("GET", _PROFILES_PATH)
        return _find_current(response.json())

    async def create(self, params: CreateProfileRequest) -> AuthenticateResponse:
        """Sign up: create a new user, organization, and profile via POST /v2/profiles.

        This is the self-service registration path, so the deployed endpoint
        requires an UNAUTHENTICATED caller (authenticated users get a 400
        pointing at POST /v2/organizations). The new profile becomes current
        and session tokens are returned, but the session is only partial
        until the emailed verification code is submitted: apply the returned
        access token with set_token(), then call auth.verify_email() on the
        same endpoint. Mirrors js-sdk createProfile.

        Example:
            tokens = await endpoint.profiles.create(CreateProfileRequest(
                email="a@b.com", password="secret!A1", first_name="First",
                last_name="Last", org_name="NEW ORG",
            ))
            endpoint.set_token(tokens.access_token)
            verified = await endpoint.auth.verify_email(email="a@b.com", token="EMAILED-CODE")

        Args:
            params: The signup fields; org_name does not need to be unique.

        Returns:
            Tokens for the new (partial, unverified) session; apply with set_token().

        Raises:
            VerdocsAPIError: The email already exists, the password is too
                weak, or the caller was already authenticated.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", _PROFILES_PATH, json=_write_body(params))
        return AuthenticateResponse.model_validate(response.json())

    async def switch(self, profile_id: str) -> AuthenticateResponse:
        """Switch the caller's current profile via POST /v2/profiles/{profile_id}/switch.

        The current profile drives permissions checking and the profile_id
        stamped on most operations, so switch before acting as another
        organization. Like the js-sdk, the new tokens are returned, not
        applied: call set_token() with the new access token to start
        operating as the switched profile (the old token keeps its previous
        profile context until it expires). Mirrors js-sdk switchProfile.

        Example:
            tokens = await endpoint.profiles.switch("PROFILE-ID")
            endpoint.set_token(tokens.access_token)

        Args:
            profile_id: ID of the caller's profile to make current.

        Returns:
            Tokens carrying the switched profile; apply with set_token().

        Raises:
            NotFoundError: The caller has no profile with that ID.
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("POST", f"{_PROFILES_PATH}/{profile_id}/switch")
        return AuthenticateResponse.model_validate(response.json())

    async def update(self, profile_id: str, params: UpdateProfileRequest) -> Profile:
        """Update a profile via PATCH /v2/profiles/{profile_id}.

        For your own profile this updates the basic fields (name, phone,
        timezone, locale); admins may also update other members of their
        organization, which is the only path that accepts permissions and
        roles (see UpdateProfileRequest for the split). Mirrors js-sdk
        updateProfile.

        Args:
            profile_id: ID of the profile to update.
            params: The fields to change; unset fields are left alone.

        Returns:
            The updated profile.

        Raises:
            NotFoundError: No profile has that ID.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        response = await self._endpoint._request("PATCH", f"{_PROFILES_PATH}/{profile_id}", json=_write_body(params))
        return Profile.model_validate(response.json())

    async def update_photo(self, profile_id: str, picture: FileInput) -> Profile:
        """Update the caller's profile photo via a multipart PATCH /v2/profiles/{profile_id}.

        Only the caller's own profile accepts a photo. The file goes up as a
        `picture` part, and the declared content type is stored with it.
        Mirrors js-sdk updateProfilePhoto, minus its browser-only upload
        progress callback.

        Args:
            profile_id: ID of the caller's own profile.
            picture: The photo: a filesystem path (str or PathLike), raw
                bytes, a binary file-like object, or an httpx-style
                (filename, content[, content_type]) tuple.

        Returns:
            The updated profile; its picture field is a public CDN URL.

        Raises:
            AuthenticationError: The endpoint has no valid user session.
            VerdocsAPIError: The API returned another non-2xx status.
            VerdocsConnectionError: The request never reached the API.
        """
        path = f"{_PROFILES_PATH}/{profile_id}"
        # _request only speaks JSON, so the multipart upload goes through the
        # raw client with the same error mapping.
        try:
            response = await self._endpoint._client.request(
                "PATCH", path, files={"picture": _picture_part(picture)}, timeout=_PHOTO_TIMEOUT
            )
        except httpx.TransportError as exc:
            raise VerdocsConnectionError(f"PATCH {path} failed: {exc}") from exc
        if response.status_code >= 400:
            raise api_error_from_response(response)
        return Profile.model_validate(response.json())

    async def delete(self, profile_id: str) -> AuthenticateResponse | None:
        """Delete one of the caller's profiles via DELETE /v2/profiles/{profile_id}.

        Deleting your current profile switches you to the next available one
        and returns its tokens (apply them with set_token()); deleting your
        last remaining profile returns None, and you are logged out. Mirrors
        js-sdk deleteProfile.

        As deployed, the handler only sends a response when the deleted
        profile was the caller's current one: deleting a NON-current profile
        succeeds but never answers, so the call times out. Switch to a
        profile before deleting it.

        Args:
            profile_id: ID of the caller's profile to delete.

        Returns:
            Tokens for the next available profile, or None when the last
            profile was deleted and the session is gone.

        Raises:
            VerdocsAPIError: The profile does not belong to the caller.
            VerdocsConnectionError: The request never reached the API, or
                (deployed quirk above) the API never answered.
        """
        response = await self._endpoint._request("DELETE", f"{_PROFILES_PATH}/{profile_id}")
        return _delete_result(response.json())

File: sdks/csharp/src/Verdocs.Sdk/Resources/Profiles.cs
using System.Net.Http.Headers;
using Verdocs.Models;

namespace Verdocs.Resources;

/// <summary>
/// Profile calls, reached through <see cref="VerdocsEndpoint.Profiles"/>.
/// </summary>
public sealed class Profiles
{
    private readonly VerdocsEndpoint _endpoint;

    internal Profiles(VerdocsEndpoint endpoint)
    {
        _endpoint = endpoint;
    }

    /// <summary>
    /// Gets the caller's profiles. A user has one profile per organization membership;
    /// exactly one is marked <see cref="Profile.Current"/> at a time and operations are
    /// performed as that profile.
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The caller's profiles.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the session is invalid.</exception>
    public Task<IReadOnlyList<Profile>> ListAsync(CancellationToken cancellationToken = default)
    {
        return _endpoint.SendAsync<IReadOnlyList<Profile>>(HttpMethod.Get, "/v2/profiles", null, cancellationToken);
    }

    /// <summary>
    /// Gets the caller's current profile. A user has one profile per organization membership
    /// and exactly one is current at a time; operations are performed as that profile.
    /// </summary>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The profile marked current, or null if the caller has none.</returns>
    /// <exception cref="VerdocsApiException">The call failed, for example because the session is invalid.</exception>
    public async Task<Profile?> GetCurrentAsync(CancellationToken cancellationToken = default)
    {
        var profiles = await _endpoint.SendAsync<List<Profile>>(HttpMethod.Get, "/v2/profiles", null, cancellationToken)
            .ConfigureAwait(false);
        return profiles.Find(profile => profile.Current);
    }

    /// <summary>
    /// Registers a new user account plus a new organization the caller will own. Call this
    /// on an unauthenticated endpoint: the deployed API rejects the request when a session
    /// is present (users join existing organizations by invitation instead). The new profile
    /// becomes the user's current profile and session tokens are returned, but the email is
    /// not verified yet, so most other calls will fail until the user submits the emailed
    /// code via <see cref="Auth.VerifyEmailAsync"/>.
    ///
    /// <example>
    /// <code>
    /// var session = await endpoint.Profiles.CreateAsync(new CreateProfileRequest
    /// {
    ///     Email = "you@example.com",
    ///     Password = "PASSWORD",
    ///     FirstName = "First",
    ///     LastName = "Last",
    ///     OrgName = "New Org",
    /// });
    /// endpoint.SetToken(session.AccessToken);
    /// // The user receives a code by email:
    /// await endpoint.Auth.VerifyEmailAsync(new VerifyEmailRequest { Email = "you@example.com", Token = code });
    /// </code>
    /// </example>
    /// </summary>
    /// <param name="request">The account and organization details.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>Session tokens for the new (not yet verified) profile.</returns>
    /// <exception cref="VerdocsApiException">The email is already registered, the caller was authenticated, or the call failed.</exception>
    public Task<AuthenticateResponse> CreateAsync(CreateProfileRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<AuthenticateResponse>(HttpMethod.Post, "/v2/profiles", request, cancellationToken);
    }

    /// <summary>
    /// Switches the caller's current profile and returns fresh session tokens for it. The
    /// current profile drives permission checks and record ownership for most operations, so
    /// select the appropriate profile before making other calls. The endpoint does not apply
    /// the new tokens automatically; pass the access token to
    /// <see cref="VerdocsEndpoint.SetToken"/>, matching <see cref="Auth.AuthenticateAsync"/>.
    /// </summary>
    /// <param name="profileId">The profile to make current. Must belong to the caller.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>Fresh session tokens for the newly current profile.</returns>
    /// <exception cref="VerdocsApiException">The profile was not found or the call failed.</exception>
    public Task<AuthenticateResponse> SwitchAsync(string profileId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(profileId);
        return _endpoint.SendAsync<AuthenticateResponse>(
            HttpMethod.Post,
            "/v2/profiles/" + Uri.EscapeDataString(profileId) + "/switch",
            null,
            cancellationToken);
    }

    /// <summary>
    /// Updates a profile: the caller's own current profile, or, for admins, another member
    /// of the same organization. The two paths accept different fields; see
    /// <see cref="UpdateProfileRequest"/>.
    /// </summary>
    /// <param name="profileId">The profile to update.</param>
    /// <param name="request">The fields to change. Unset properties are left as they are.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The updated profile.</returns>
    /// <exception cref="VerdocsApiException">The profile was not found, a field was not accepted, or the call failed.</exception>
    public Task<Profile> UpdateAsync(string profileId, UpdateProfileRequest request, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(profileId);
        ArgumentNullException.ThrowIfNull(request);
        return _endpoint.SendAsync<Profile>(
            HttpMethod.Patch,
            "/v2/profiles/" + Uri.EscapeDataString(profileId),
            request,
            cancellationToken);
    }

    /// <summary>
    /// Deletes one of the caller's profiles. Deleting the current profile switches the
    /// caller to the next available one and returns fresh tokens for it; deleting the last
    /// profile logs the caller out and returns a status message instead (see
    /// <see cref="DeleteProfileResponse"/>). Deleting a profile that is NOT current performs
    /// the delete but the deployed API sends no response at all, so the call ends in a
    /// TimeoutException. Until that is fixed server-side, the workaround is to
    /// <see cref="SwitchAsync"/> to the profile first and then delete it.
    /// </summary>
    /// <param name="profileId">The profile to delete. Must belong to the caller.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>Tokens for the next profile, or a logout notice when it was the last one.</returns>
    /// <exception cref="VerdocsApiException">The profile was not found or the call failed.</exception>
    public Task<DeleteProfileResponse> DeleteAsync(string profileId, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(profileId);
        return _endpoint.SendAsync<DeleteProfileResponse>(
            HttpMethod.Delete,
            "/v2/profiles/" + Uri.EscapeDataString(profileId),
            null,
            cancellationToken);
    }

    /// <summary>
    /// Replaces the profile photo. Only the caller's own current profile accepts a photo.
    /// The stream is read to the end and disposed when the request completes.
    /// </summary>
    /// <param name="profileId">The profile to update. Must be the caller's current profile.</param>
    /// <param name="photo">The image content.</param>
    /// <param name="fileName">Filename to report for the upload, for example "avatar.png".</param>
    /// <param name="contentType">MIME type of the image, for example "image/png". The server stores and serves the declared type without validating it.</param>
    /// <param name="cancellationToken">Token to cancel the operation.</param>
    /// <returns>The updated profile; <see cref="Profile.Picture"/> carries the new photo URL.</returns>
    /// <exception cref="VerdocsApiException">The upload was rejected or the call failed.</exception>
    public Task<Profile> UpdatePhotoAsync(string profileId, Stream photo, string fileName, string contentType, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrEmpty(profileId);
        ArgumentNullException.ThrowIfNull(photo);
        ArgumentException.ThrowIfNullOrEmpty(fileName);
        ArgumentException.ThrowIfNullOrEmpty(contentType);
        if (!MediaTypeHeaderValue.TryParse(contentType, out var mediaType))
        {
            throw new ArgumentException("The content type must be a valid MIME type, for example \"image/png\".", nameof(contentType));
        }

        // The handler only looks at the file part named "picture" (sdks/WIRE-NOTES.md); parts
        // under any other name are silently ignored. The transport disposes the request and
        // with it this content tree, including the caller's stream.
        var content = new MultipartFormDataContent();
        var picture = new StreamContent(photo);
        picture.Headers.ContentType = mediaType;
        content.Add(picture, "picture", fileName);

        return _endpoint.SendAsync<Profile>(
            HttpMethod.Patch,
            "/v2/profiles/" + Uri.EscapeDataString(profileId),
            content,
            cancellationToken);
    }
}
